### PR TITLE
[QNN] fix Clip and ReLU removal logic

### PR DESCRIPTION
**Pull request by cloudhan**
Depends on #22580, see the PR for more details.

### CODE REVIEW

**Comment by adrianlizarraga** [comment]
Hi @cloudhan, Thanks for this PR! The Windows x64 QNN CI Pipeline seems to be failing. Could be related to these updates.

**Comment by cloudhan** [comment]
> The Windows x64 QNN CI Pipeline seems to be failing. Could be related to these updates.

@adrianlizarraga  How should I understand a still open PR to cause CI pipeline to fail?

BTW, do we need additional tests for this PR, in case it is not covered in the existing tests. Also please take a look.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc [hunk 1]
```diff
@@ -17,6 +18,50 @@
...
+//      upperbound of min and lowerbound of max
+//    - Why?
+//      We want the codomain to be unchanged, so as long as the domain genreate a codomain that fill the integer value
+//      range will be fine.
+
+// The quantization formula is y = saturate((x / y_scale) + y_zero_point)
+// For (x / y_scale), it rounds to the nearest even. So the allowed quantize limits before saturate need to be taken
+// care of.
+//
+// The following struct provides a wrapper to compute the domain from codomain (Q output dtype).
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:39]: Thank you for the clear explanation.

**Review by adrianlizarraga — APPROVED** [review]

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc [hunk 2]
```diff
@@ -115,15 +160,14 @@ static bool CanClipBeRemoved(const QnnModelWrapper& qnn_model_wrapper,
...
-  // The clip range must entirely overlap the quantization range (quantization can be smaller).
-  // Clip range:   [------------------]
-  // Quant range:    [-------------]
-  constexpr float epsilon = std::numeric_limits<float>::epsilon();
-  if ((epsilon < clip_min - rmin) || (epsilon < rmax - clip_max)) {
-    return false;
+  // The Clip codomain must entirely overlap the QuantizeLinear domain (quantization can be smaller).
+  // Clip codomain:          [------------------]
+  // QuantizeLinear domain:    [-------------]
+  if (clip_min <= min_ && max_ <= clip_max) {
```
Comment (adrianlizarraga) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:166]: Can we add a unit test for the case this is fixing? Here's an example unit test that builds a model (programmatically) and then calls the `RunQnnModelTest` function to check that the output from QNN matches CPU EP.

https://github.com/microsoft/onnxruntime/blob/1fa219d7d517314ad51bc211702539714ddcbf8a/onnxruntime/test/providers/qnn/qnn_basic_test.cc#L970

Comment (cloudhan) [reply onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:166]: I tried to construct a graph with 
```
input --> Add -> Clip-> QuantizeLinear-> output
    Zeros  /
```
and send carefully constructed inputs around boundary.

The problem here is that we cannot actually put these carefully constructed graph on the QNN EP as a whole (at least for `Clip-> QuantizeLinear` part).

So I am not sure if we can actually achieve this type of testing...

https://github.com/microsoft/onnxruntime/blob/bc24d3b76609b9b1631fb256fb81025a6e0ee793/onnxruntime/test/providers/qnn/conv_test.cc#L895 and https://github.com/microsoft/onnxruntime/blob/bc24d3b76609b9b1631fb256fb81025a6e0ee793/onnxruntime/test/providers/qnn/conv_test.cc#L978
have already covered the fusion part, but not the boundary part.

File: onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc [hunk 3]
```diff
@@ -53,11 +54,11 @@ static bool GetQScalarScaleZeroPoint(const QnnModelWrapper& qnn_model_wrapper,
   return true;
 }
 
-// Computes the floating point range (rmin, rmax) from a QuantizeLinear node's scale/zero-point.
-static bool GetQRminRmax(const QnnModelWrapper& qnn_model_wrapper,
-                         const NodeUnit& q_node_unit,
-                         /*out*/ float& rmin,
-                         /*out*/ float& rmax) {
+// Computes the floating point domain (min_, max_) from a QuantizeLinear node's scale/zero-point.
+static bool GetQDomain(const QnnModelWrapper& qnn_model_wrapper,
```
Comment (skottmckay) [inline onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:58]: Is it possible to have a shared implementation of the logic for this and #22580 to use?

Comment (cloudhan) [reply onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:58]: I can refactor the `GetQDomain` in [onnxruntime/core/optimizer/qdq_transformer/clip_quantizelinear.cc](https://github.com/microsoft/onnxruntime/pull/22580/files?w=1#diff-d263e4275790b1e6e406f2513688257815e385230bc3a22f11566f611e6778b8) in this PR and reuse it here.

Comment (cloudhan) [reply onnxruntime/core/providers/qnn/builder/qnn_node_group/conv_activation_fusion.cc:58]: Seems only the `switch` part can be reused, which only improves very little. Lets just keep it as is if no future reuse pop up.

**Pull request closed without merging**